app/plugins/weasyl.ts

```diff
@@ -11,41 +12,128 @@ export class WeasylPlugin extends BaseSitePlugin {
+        // or bail out if not found.
+        //
+        // When Raccoony is trying to decide if it should run on a
+        // particular page, it calls getMedia() first, and then
+        // getPageLinkList().  So getMedia() will be called on pages
+        // without a canonical link, such as the "recently popular"
+        // page.  Detect that situation and bail out, so Raccoony can
+        // go on and call getPageLinkList().
+        let canonical = document.querySelector("link[rel=canonical]");
+        if (!canonical) {
+            return Promise.resolve(null);
+        }
+
+        // Get the native "download" button, or bail out if not found.
+        //
+        // Some pages have a "canonical" link, but no download button,
+        // such as a user's main profile page.  Detect that situation
+        // and bail out, so we don't try to parse things that may not
+        // be there.
         let buttonSpan = document.querySelector("#detail-actions .icon-arrowDown");
         if (!buttonSpan) {
             return Promise.resolve(null);
         }
         let button = buttonSpan.parentElement;
 
-        // Get the URL
-        // Weasyl download URLs are of the format 
-        // https://cdn.weasyl.com/~[user]/submissions/[id]/[hash]/[user]-[filename.ext]?download
-        let url = button.getAttribute("href");
-        let urlObj = new URL(url);
+        // If we get here, we probably have a canonical link and a download
+        // button.
+
+        // The canonical link has at least two formats; handle them both.
+        // "visual", "literary", and "multimedia" submissions:
+        // https://www.weasyl.com/%7E[user]/submissions/[id]/[title]
+        // "character" submissions:
+        // https://www.weasyl.com/character/[id]/[title]
+        let canonicalLink = canonical.getAttribute("href");
+        logger.log("weasyl: canonical href ", canonicalLink);
+
+        // Split on slashes
+        let canonicalObj = new URL(canonicalLink);
+        let pathParts = canonicalObj.pathname.split('/');
 
-        // Get the username
-        let pathParts = urlObj.pathname.split('/'); // "/~user/..."
-        let username = pathParts[1] || "";
-        username = username.replace("~", "");
+        let username = "";
+        let submissionIDIndex = 0;
+        let filenameIndex = 0;
 
-        // Get the filename
-        let serviceFilename = pathParts.pop();
-        pathParts.pop(); //pop off the unneeded hash
-        let id = pathParts.pop();
+        if(canonicalLink.indexOf("/submissions/") > -1) {
+            // regular submission
+            logger.log("weasyl: regular submission");
 
-        // Strip off the username
-        let { filename, ext } = getFilenameParts(serviceFilename);
-        filename = filename.substring(username.length + 1);
+            // Get the username and kill leading '%7E'
+            username = pathParts[1] || "";
+            username = username.slice(3);
```

Comment:
Probably the excruciatingly correct way to do this would be to urldecode and then strip off the leading tilde:
```typescript
username = decodeURIComponent(username).replace(/^~/, "");
```

Reply:
I did feel like the "this is a URL" processing should have already happened, and I would have just seen the '~' instead of the '%7E' by the time I got it, but it didn't work that way.  I agree that treating it as a URL and modifying it that way is probably a better idea.  If you want to change it to your proposal, go ahead.